Kernel/Devices/VMWareSVGA/VMWareSVGADevice.cpp

```diff
@@ -0,0 +1,286 @@
+
+void VMWareSVGADevice::initialize()
+{
+    s_the.ensure_instance();
+}
+
+VMWareSVGADevice& VMWareSVGADevice::the()
+{
+    return *s_the;
+}
+
+// We put a fake IRQ into the IRQ Handler, as we will change it anyway
+// TODO: this is a really nasty hack that only works as the IRQ is registered on the first call to enable_irq()
+// This should probably be passed in by the init_stage2
+VMWareSVGADevice::VMWareSVGADevice()
+    : BlockDevice(29, 0)
+    , IRQHandler(0)
+{
+    dbgln("VMWareSVGA: init");
+    // Find the VMWare graphics
+    static const PCI::ID vmware_id = { 0x15AD, 0x0405 };
+    PCI::enumerate([&](const PCI::Address& address, PCI::ID id) {
+        if (id == vmware_id) {
+            // Allocate some sheeet
+            m_io_base = PCI::get_BAR0(address) & 0xfffffff0;
+            m_framebuffer_address_physical = PhysicalAddress(page_base_of(PCI::get_BAR1(address)));
+            m_framebuffer_max_size = PCI::get_BAR_space_size(address, 1);
+            m_fifo_mapping = MM.allocate_kernel_region(PhysicalAddress(page_base_of(PCI::get_BAR2(address) & 0xfffffff0)), page_round_up(PCI::get_BAR_space_size(address, 2)), "VMWareSVGA Driver FIFO", Region::Access::Read | Region::Access::Write, Region::Cacheable::No);
+            m_fifo_addr = m_fifo_mapping->vaddr();
+            m_interrupt_line = PCI::get_interrupt_line(address);
+            change_irq_number(m_interrupt_line);
+            dbgln("VMWareSVGA: pci address: {}, io_base: {:x}, framebuffer phys: {:x}", address, m_io_base, m_framebuffer_address_physical.as_ptr());
+            // Clear framebuffer memory with arbitrary value
+            OwnPtr<Region> test_map = MM.allocate_kernel_region(m_framebuffer_address_physical, page_round_up(PCI::get_BAR_space_size(address, 1)), "VMWareSVGA FB Test", Region::Access::Write, Region::Cacheable::No);
+            // memset(test_map->vaddr().as_ptr(), 0xCC, PCI::get_BAR_space_size(address, 1));
+            test_map->unmap();
+        }
+    });
+    VERIFY(m_io_base);
+    VERIFY(m_framebuffer_address_physical.as_ptr());
+    VERIFY(m_fifo_addr.as_ptr());
+
+    // Determine SVGA Version
+    // The way this is done ist by writing the highest version we support (2) into a device support register,
+    // and checking if the value has been changed.
+    // If it hasnt, then we found the highest version that both we and the "card" support.
+    // If it has, decrement by one and repeat.
+    m_device_version = make_version_id(2);
+    set_register(IORegister::ID, m_device_version);
+    while (get_register(IORegister::ID) != m_device_version && m_device_version >= make_version_id(0)) {
+        m_device_version--;
+        set_register(IORegister::ID, m_device_version);
+    }
+
+    VERIFY(m_device_version >= make_version_id(0));
+
+    // We want to read the size of the Framebuffer
+    m_fifo_size = get_register(IORegister::MemSize);
+    m_framebuffer_fb_size = get_register(IORegister::MemSize);
```

Comment:
Are these supposed to be read from the same register or is that a typo?

Reply:
That is a typo, but m_framebuffer_fb_size should never be used as it returns the same value as getting the size via PCI::get_bar_size
The fifo size we need, and reads from the correct register